### PR TITLE
core/state: add account address to Trie slot accessors

### PR DESCRIPTION
**Pull request by gballet**
A new proposal to make the `Trie` interface less MPT-dependent.

After the introduction of the `TryAccount*` functions, `TryGet`, `TryUpdate` and `TryDelete` are now only meant to read an account's storage. In their current form, they assume that an account storage is stored in a separate trie, and that the hashing of the slot is independent of its account's address.

The proposed structure for a stateless storage breaks these two assumptions: the hashing of a slot key requires the address and all slots and accounts are stored in a single trie.

This PR therefore adds an `address` parameter to the interface. It is ignored in the MPT version, so this change has no functional impact, however it will reduce the diff size when merging verkle trees.

Since in their current implementation, verkle trees don't use the prefetcher, I considered not touching it and using `nil`. I decided to proceed anyway for consistency's sake.

### CODE REVIEW

**Comment by rjl493456442** [comment]
I think we can be even more aggressive by changing functions to `TryGetStorage(common.Address, []byte)`, `TryUpdateStorage(common.Address, []byte)` and `TryDeleteStorage(common.Address, []byte)`. 

Originally, `TryGet`, `TryUpdate` and `TryDelete` is the generic version for both accounts and storages. Now we can have the specific version for accounts and storages respectively which is the cleaner solution to me.

File: core/state/trie_prefetcher.go [hunk 1]
```diff
@@ -336,7 +338,7 @@ func (sf *subfetcher) loop() {
 					if _, ok := sf.seen[string(task)]; ok {
 						sf.dups++
 					} else {
-						sf.trie.TryGet(task)
+						sf.trie.TryGet(sf.addr, task)
```
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:341]: We can determine the task type by length, e.g.

```go
if len(task) == common.AddressLength {
	sf.trie.TryGetAccount(common.BytesToAddress(task))
} else {
	sf.trie.TryGetStorage(sf.address, task)
}
sf.seen[string(task)] = struct{}{}
```

File: core/state/database.go [hunk 2]
```diff
@@ -68,10 +68,10 @@ type Trie interface {
...
 	GetKey([]byte) []byte
 
-	// TryGet returns the value for key stored in the trie. The value bytes must
-	// not be modified by the caller. If a node was not found in the database, a
-	// trie.MissingNodeError is returned.
-	TryGet(key []byte) ([]byte, error)
+	// TryGetStorage returns the value for key stored in the trie. The value bytes
+	// must not be modified by the caller. If a node was not found in the database,
+	// a trie.MissingNodeError is returned.
+	TryGetStorage(addr, key []byte) ([]byte, error)
```
Comment (rjl493456442) [inline core/state/database.go:74]: Please change the address type to `common.Address`. We did the same thing for `TryGetAccount`.

**Review by rjl493456442 — APPROVED** [review]
lgtm

**Review by fjl — APPROVED** [review]
We should try a benchmark run and check the trie prefetcher metrics.

**Review by holiman — APPROVED** [review]
LGTM
After merging, we should swap this into a running benchmarker node, and check the meters, to ensure that the performance is not broken:
```
		deliveryMissMeter: metrics.GetOrRegisterMeter(prefix+"/deliverymiss", nil),
		accountLoadMeter:  metrics.GetOrRegisterMeter(prefix+"/account/load", nil),
		accountDupMeter:   metrics.GetOrRegisterMeter(prefix+"/account/dup", nil),
		accountSkipMeter:  metrics.GetOrRegisterMeter(prefix+"/account/skip", nil),
		accountWasteMeter: metrics.GetOrRegisterMeter(prefix+"/account/waste", nil),
		storageLoadMeter:  metrics.GetOrRegisterMeter(prefix+"/storage/load", nil),
		storageDupMeter:   metrics.GetOrRegisterMeter(prefix+"/storage/dup", nil),
		storageSkipMeter:  metrics.GetOrRegisterMeter(prefix+"/storage/skip", nil),
		storageWasteMeter: metrics.GetOrRegisterMeter(prefix+"/storage/waste", nil),

```

**Comment by holiman** [comment]
I'll deploy this on `bench07`

**Comment by holiman** [comment]
All good. If this had been wrong, it would be evident on the charts. But there's no impact: 

Screenshot 2023-03-23 at 11-58-45 Dual Geth - Grafana

**Pull request merged**